app/src/main/java/org/mozilla/focus/widget/EqualDistributeGrid.java

```diff
@@ -0,0 +1,178 @@
+        final int rowLength = Math.min(rowCapacity, visibleCnt);
+
+        // to know width and height for each cell
+        final View reference = getFirstNotGoneChild();
+
+        int width = calculateWidth(widthMeasureSpec, heightMeasureSpec, visibleCnt, rowLength, reference);
+        int height = calculateHeight(widthMeasureSpec, heightMeasureSpec, visibleCnt, rowLength, reference);
+        setMeasuredDimension(width, height);
+    }
+
+    private int calculateWidth(int widthMeasureSpec,
+                               int heightMeasureSpec,
+                               int visibleCnt,
+                               int rowLength,
+                               @Nullable View reference) {
+        final int widthMode = MeasureSpec.getMode(widthMeasureSpec);
+        final int widthSize = MeasureSpec.getSize(widthMeasureSpec);
+
+        if (widthMode == MeasureSpec.EXACTLY) {
+            return widthSize;
+        }
+
+        int width = (reference == null || visibleCnt == 0)
+                ? 0
+                : reference.getMeasuredWidth() * rowLength;
+
+        if (widthMode == MeasureSpec.AT_MOST) {
+            width = Math.min(width, widthSize);
+        }
+
+        return width;
+    }
+
+    private int calculateHeight(int widthMeasureSpec,
+                                int heightMeasureSpec,
+                                int visibleCnt,
+                                int rowLength,
+                                @Nullable View reference) {
+        final int heightMode = MeasureSpec.getMode(heightMeasureSpec);
+        final int heightSize = MeasureSpec.getSize(heightMeasureSpec);
+
+        if (heightMode == MeasureSpec.EXACTLY) {
+            return heightSize;
+        }
+
+        if (rowLength == 0) {
+            return 0;
+        }
+
+        int lines = (visibleCnt / rowLength) + (visibleCnt % rowLength == 0 ? 0 : 1);
+
+        final int cellsOccupied = (reference == null || visibleCnt == 0)
+                ? 0
+                : lines * reference.getMeasuredHeight();
+
+        int height = cellsOccupied
+                + (lines - 1) * gutter
+                + getPaddingTop() + getPaddingBottom();
+
+        return height;
+    }
+
+    @Override
+    protected void onLayout(boolean changed, int l, int t, int r, int b) {
+        final int visibleCnt = getNotGoneCount();
+
+        // to know width and height for each cell
+        final View reference = getFirstNotGoneChild();
+
+        if (reference == null || visibleCnt == 0) {
+            return;
+        }
+
+        final int cellWidth = reference.getMeasuredWidth();
+        final int cellHeight = reference.getMeasuredHeight();
+        final int paddingStart = getPaddingStart();
+        final int paddingEnd = getPaddingEnd();
+
+        // how many cells in first row
+        int rowLength = Math.min(rowCapacity, visibleCnt);
+
+        // space between cells in horizontal orientation
+        int spacing = (rowLength == 1)
```

Comment:
Still not handling visibleCnt == 0? It can be simply a RuntimeException if you prefer so.

Reply:
if `visibleCnt == 0`, it will return at line 122

Reply:
Ah right sorry for missing that!